airflow/sentry.py

```diff
@@ -0,0 +1,203 @@
+    """
+    Retrieves task instance based on dag_id and execution_date.
+    """
+    from airflow.models.taskinstance import TaskInstance  # Avoid circular import
+
+    if session is None or not task_ids:
+        return []
+
+    TI = TaskInstance
+    ti = (
+        session.query(TI)
+        .filter(
+            TI.dag_id == dag_id,
+            TI.task_id.in_(task_ids),
+            TI.execution_date == execution_date,
+        )
+        .all()
+    )
+    return ti
+
+
+class ConfiguredSentry:
+    """
+    Configure Sentry SDK.
+    """
+
+    SCOPE_TAGS = frozenset(("task_id", "dag_id", "execution_date", "operator"))
+    SCOPE_CRUMBS = frozenset(
+        ("dag_id", "task_id", "execution_date", "state", "operator", "duration")
+    )
+
+    def __init__(self):
+        """
+        Initialize the Sentry SDK.
+        """
+        ignore_logger("airflow.task")
+        ignore_logger("airflow.jobs.backfill_job.BackfillJob")
+        executor_name = conf.get("core", "EXECUTOR")
+
+        sentry_flask = FlaskIntegration()
+
+        # LoggingIntegration is set by default.
+        integrations = [sentry_flask]
+
+        if executor_name == "CeleryExecutor":
+            from sentry_sdk.integrations.celery import CeleryIntegration
+            from sentry_sdk.integrations.tornado import TornadoIntegration
+
+            sentry_celery = CeleryIntegration()
+            sentry_tornado = TornadoIntegration()
+            integrations += [sentry_celery, sentry_tornado]
+
+        self.task = None
+        self.execution_date = None
+        self.dag_id = None
+
+        dsn = conf.get("sentry", "sentry_dsn")
+        if dsn:
+            init(dsn=dsn, integrations=integrations)
+        else:
+            # Setting up Sentry using environment variables.
+            log.debug("Defaulting to SENTRY_DSN in environment.")
+            init(integrations=integrations)
+
+    def add_tagging(self, task_instance):
+        """
+        Function to add tagging for a task_instance.
+        """
+        self.task = task_instance.task
+        self.execution_date = task_instance.execution_date
+        self.dag_id = task_instance.dag_id
+
+        with configure_scope() as scope:
+            for tag_name in self.SCOPE_TAGS:
+                attribute = getattr(task_instance, tag_name)
+                if tag_name == "operator":
+                    attribute = self.task.__class__.__name__
+                scope.set_tag(tag_name, attribute)
+
+    @provide_session
+    def add_breadcrumbs(self, session=None):
+        """
+        Function to add breadcrumbs inside of a task_instance.
+        """
+
+        if self.task is None or session is None:
+            return
+
+        dag = self.task.dag
+        task_ids = dag.task_ids
+        task_instances = get_task_instances(
+            self.dag_id, task_ids, self.execution_date, session=session
+        )
+
+        for ti in task_instances:
+            if ti is not None and (
```

Comment:
Why would `ti` be `None`? Also, if we only care about task instances with the `SUCCESS` or `FAILED` state, can we add that filter to `get_task_instances`?

Reply:
Also, the code is calling these upstream tasks, but it looks like the filter checks for all successful or failed tasks in the dag run. I think that could include tasks that aren't upstream of the relevant task instance. It might be more accurate to use something like `get_flat_relative_ids` instead.

Reply:
@jmcarp 

> Also, the code is calling these upstream tasks, but it looks like the filter checks for all successful or failed tasks in the dag run. I think that could include tasks that aren't upstream of the relevant task instance. It might be more accurate to use something like `get_flat_relative_ids` instead.

This is what this PR did before but we changed it to include all tasks in the dag run, as this could give context if an error occurred due to a race condition with a parallel task,

Reply:
OK, then maybe we should keep the behavior but use a clearer name--how about `completed_tasks`?

Reply:
I changed the name to reflect that they are completed tasks.